### DISCUSSION TITLE
👋 Who is using Nodify?

### DISCUSSION THREAD

**Discussion by miroiu** [miroiu/nodify#158] [Show and tell]
I would love to see what you’re building with Nodify! Whether it’s a public app, a private tool, or even just a demo, feel free to share it here.

If possible, please include:
* Project name/Organization name
* Description
* Screenshot
* Link to the website, store or GitHub repository

**Comment by PyPete** [comment]
Hello,
I use Nodify in a private application which I hope to release open source in the future.

Title: Process Orchestration Platform (POP)
Description: Lab Automation framework that orchestrates individual devices in a science laboratory environment to achieve complex chemistry workflows. More information see [https://www.youtube.com/watch?v=RTjri9tRjLo&list=LLdhjLwwgEzijbZciAIH-Weg](url) (starts around 19 mins)
Screenshots of the Nodify part showing the workflow designer:
![image](https://user-images.githubusercontent.com/12038586/229513682-6522e9c0-e701-49a6-b75e-c26f83da2d76.png)

![image](https://user-images.githubusercontent.com/12038586/229513752-d28cf3a2-e432-475c-80bf-6f5e868276f7.png)

**Comment by ApokalmiX** [comment]
Hi,
I use Nodify in a private project named HiysLight.
The goal of the project is to create an environment linking audio inputs and MIDI events to the control of DMX lights.
Nodify is used for the creation of macro allowing the interpretation of the audio inputs towards the management of the lights.
The project is not yet open source, maybe I'll put it when I make a stable and more flexible version.

![image](https://user-images.githubusercontent.com/16222688/230422834-8913fc9b-3db9-4cb1-905e-96f168930df4.png)

I take this opportunity to congratulate you on the quality of Nodify, it is one of the rare entirely MVVM project, which does not impose any constraints and which is fully customizable that I have been able to integrate. I will use it again without hesitation when I need it!

**Comment by trrahul** [comment]
Hey, I had ported Nodify to Avalonia a while back and made it to private. I have now made it open and available here, https://github.com/trrahul/nodify-avalonia. 

I also would like to mention that this library is being used in the development of an indie game, and I cannot share any further details. I haven't had an opportunity to polish the code and examples and include the new features that have been added to Nodify. However, there is one change I made to update the backend real-time when the user moves a rendered point. I added the PreviewLocation property to the node. 

I would greatly appreciate any feedback and, more importantly, any contributions made towards enhancing the port. 

Thank you @miroiu for your amazing work.

**Comment by trrahul** [comment]
Also, I have used the port in one of my private projects. I have a demo video of one of the initial versions here. https://www.youtube.com/watch?v=vJu3wMTGsGU

![image](https://github.com/miroiu/nodify/assets/7353840/d7311284-91e9-47f8-a1bd-d80f7c6d77ab)

**Comment by miroiu** [comment]
Hi @trrahul ! Thanks a bunch for taking the time to port Nodify to Avalonia and sharing it. It's awesome to see how people are getting creative with Nodify.

I personally don't have much experience with Avalonia, so I'm not sure how much help I can offer in terms of contributions to your repository, and we've actually got a [recommended port](https://github.com/BAndysc/nodify-avalonia) now that's closer to the original, but I really appreciate the effort you put into your version.

Best of luck with your private project! Your idea definitely shows promise. Keep up the great work! And if you have any questions or need assistance with anything else, feel free to reach out!

**Comment by jpk6789** [comment]
Hey, after some months of on and off work on it, I can finally share the current state of work: I am working on implementing Nodify in [DMXControl 3](https://dmxcontrol.de/en/), one of the largest freeware light control projects for stage and club lighting released by the German non-profit organization [DMXControl Projects e.V.](https://dmxcontrol-projects.org/en/)

Currently we have an XNA based self-designed node editor for our so called Input Assignment. In there, users can route inputs and outputs e.g. from/to external control devices like MIDI consoles, Stream Decks, keyboard, etc. to internal functions (e.g. to control masters or trigger scenelists). Also mappings between internal functions are possible (so nowadays the name Input Assignment is a bit misleading 😂).

However, we need to throw out XNA because it is EOL and very ugly (in the image, the Stage View also uses XNA to see the difference). Nodify is a perfect replacement for the current node editor because it not only brings WPF and MVVM with it but looks so much better and is so much easier to control than our old one. The current state of the implementation works already quite well in my internal version despite the fact that I have not implemented all functionalities yet. So I will continue working on it and hope we can release the Nodify-based Input Assignment with v3.3.1 or v3.3.2 of DMXControl 3.  Here a glimpse into the current state:

![grafik](https://github.com/user-attachments/assets/f1d96f4b-bfc2-40e8-9c79-caa8ef562d41)

I am even thinking about if it is possible to replace the Stage View where we visualize the current control state of the light devices with a (maybe modified) Nodify instance as well. But I have not put any further thoughts into that yet. 

So I want to thank you @miroiu for this amazing piece of software and your hard work on it :blush:

**Comment by miroiu** [comment]
Thank you so much for sharing the progress and your kind words! I'm very happy to see Nodify being used in all these amazing projects 😊. 

The subtle customization you did by changing the border color of the nodes blends perfectly with the overall design of the application. It's exciting to hear about the potential of using Nodify for the Stage View as well. 

Looking forward to seeing further updates and the final implementation. 

Keep up the great work! 🚀

**Comment by bhxxwx** [comment]
Hello, I use Nodify in a private application .
Title: CAN data visualization
Description: A simple CAN data visualization program designed for our lab.
![image](https://github.com/user-attachments/assets/25748408-e319-4081-a4ec-d6f50714b406)

**Reply by kindergartenlaoda** [reply]
If it's possible, could you kindly make the source code public?

**Comment by miroiu** [comment]
That's cool 😎. Thanks for sharing! I see a combination of the Playground and Calculator apps, including StringMath.

**Comment by miroiu** [comment]
Developed by [xieyan110](https://github.com/xieyan110), AI Robot is a versatile node-based application designed for automation across multiple platforms and devices. Here’s what it offers:
 - Mouse and Keyboard Control: The application allows users to manipulate the mouse and keyboard through node-based interactions.
 - Android Device Automation: Utilizes ADB (Android Debug Bridge) to control Android smartphones remotely.
 - Workflow Management: Workflows can be created and saved in JSON format.
 - Workflow Execution: The home interface enables loading workflow lists and executing multiple workflows concurrently.

🚀 Check it out here: [AI Robot GitHub Repository](https://github.com/xieyan110/AIbotot)

![workflow](https://github.com/user-attachments/assets/46cdbf80-5c6f-4e7f-8e69-3544b6f1276e)

Issue reference: #157

**Comment by morning-STARK** [comment]
We are currently using Nodify in a solar cell inspection project that involves a machine vision platform with agile drag-and-drop capabilities. The system is built around powerful vision libraries like Halcon and OpenCV.
As customer demands for higher accuracy, speed, and maintainability continue to grow, traditional hardcoded vision pipelines have become difficult to adapt and debug. To address this, we modularized our vision algorithms into reusable nodes and used Nodify to create a visual node-based editor.
Now, developers and field engineers can quickly build or adjust inspection workflows by simply dragging and connecting nodes—without diving deep into the underlying code. This approach has significantly improved development efficiency and reduced the cost of maintenance and customization.
Thanks to Nodify’s MVVM architecture and high flexibility, we were able to seamlessly integrate Halcon’s HObject viewer, parameter configuration panels, and real-time debugging tools. It has truly enabled us to achieve visual algorithm design, configurable workflows, and traceable debugging.
Big thanks to the Nodify team for this excellent framework—it has made vision system development more intuitive, agile, and sustainable!
>This message was translated by AI—please excuse any awkward phrasing or errors.

目前正在一个太阳能电池片视觉检测项目中使用 Nodify，这是一个面向机器视觉的敏捷拖拽平台，核心集成了 Halcon 和 OpenCV 等视觉算法库。
随着客户对检测精度、速度和可维护性的要求不断提高，传统的硬编码算法流程逐渐暴露出灵活性差、调试困难的问题。为此，我们将各类视觉算法模块抽象为独立的节点，借助 Nodify 提供的节点编辑器，实现了图形化、模块化的算法流程搭建。
开发人员和现场工程师可以通过拖拽节点、连接数据流的方式快速构建或调整检测流程，无需深入代码层面。这种方式不仅大幅提升了开发效率，也显著降低了后期维护和客户需求变更的成本。
Nodify 的 MVVM 架构和灵活的自定义能力，使我们能够轻松集成 Halcon 的 HObject 查看器、参数配置界面以及实时调试模块，真正做到了“算法可视化、流程可编排、调试可追溯”。
感谢 Nodify 提供的优秀框架，让视觉系统的开发变得更加直观、敏捷、可持续！
<img width="1878" height="1010" alt="无标题" src="https://github.com/user-attachments/assets/22898bd9-c0df-4f1a-b1fd-5a1a4286a48c" />

**Reply by miroiu** [reply]
Thank you for sharing this! It's inspiring to see Nodify being used in such a meaningful project.

**Comment by GRPDAN01** [comment]
Hello,
I've implemented Nodify in our PLM (Product LifeCycle Management)
I'm using it to validate an objet change. For exemple validating the next status of an object.
I can also use it to perform actions when an event occurs.
I've created different type of nodes in a Unreal Engine style.

Here is a screenshot :

<img width="1896" height="1040" alt="Capture d&#39;écran 2026-02-03 105737" src="https://github.com/user-attachments/assets/cd50dd96-69ae-427d-a169-405a622cbb03" />


Thank you very much for your wonderful tool !

**Reply by miroiu** [reply]
Thank you very much for sharing!

**Comment by chaojian-zhang** [comment]
**Methodox Technologies, Inc.** presents **_Divooka™_** - a proper graph-first 4th-generation general-purpose high-level multi-paradigm visual programming language.

A teaser on what's coming...

<img width="1359" height="607" alt="Module Scoped Instance Node - Events Context" src="https://github.com/user-attachments/assets/59511588-b43e-41d5-ba4c-5c48d1413b2d" />

<img width="1349" height="454" alt="Module Generated Type" src="https://github.com/user-attachments/assets/3f1b688d-93cd-4259-afa4-c593ec68ca55" />

<img width="1367" height="848" alt="Node Level Automatic Dispatch (with Generic)" src="https://github.com/user-attachments/assets/fa6b41e3-b4d7-4cc4-b72b-60080e78c0af" />

<img width="2129" height="1073" alt="Flappy Bird (Basic) in Divooka" src="https://github.com/user-attachments/assets/19f4c7cf-e35f-449b-8e5a-1ea90de66841" />

<img width="1882" height="833" alt="Procedural Recursive Calls in Events Contenxt in Divooka" src="https://github.com/user-attachments/assets/37101c70-9012-4fe9-8812-20a1a77fff43" />

[Pre-Beta Annoucement](https://itch.io/t/4911014/methodox-divooka-general-purpose-visual-programming-language)

The system features two context: [dataflow](https://wiki.methodox.io/en/Divooka/Language/DataflowContext) and [procedural](https://wiki.methodox.io/en/Divooka/Language/ProceduralContext). The goal is to build not only a system but reusable components as a [C# visual programming engine](https://methodox.itch.io/divooka-engine) - though due to limited resources development on this has been slow😂 so the focus is mostly on releasing a self-contained distribution first.

Been prototyping with Nodify since 2022 - it's been a long journey and a love/hate struggle with Nodify (mostly due to limitations of MVVM) but things are smoother now! Learning to dig into source code when needed and @miroiu has been very helpful in answering questions!

More details in the future (now available on [Steam](https://store.steampowered.com/app/4422920/Divooka_Explore/))!

Websites:

* [Methodox](https://methodox.io/)
* [Divooka](https://divooka.com/)
* [Itch.io Demo](https://methodox.itch.io/divooka-explore)

**Reply by miroiu** [reply]
I originally started working on Nodify with the goal of building a visual programming system similar to Unreal Engine’s Blueprints, but integrated into the .NET ecosystem. However, I quickly realized that the scope of such a project is massive, far beyond what I could realistically tackle. I’m genuinely excited to see someone bring this idea to life. Thanks for sharing!

**Reply by chaojian-zhang** [reply]
My pleasure to stand on the shoulder of a giant!

**Comment by trrahul** [comment]
I've updated my Avalonia [port](https://github.com/trrahul/nodify-avalonia) of Nodify to be on par with the latest 7.3.0 release, porting the full feature set. It targets Avalonia 12 / .NET 8 and the library is now generally usable. Please check it out!
